tests/bluetooth/bsim_bt/bsim_test_gatt_settings/src/server.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "utils.h"
+#include "main.h"
+#include "gatt_utils.h"
+
+#include <zephyr/bluetooth/addr.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/toolchain/gcc.h>
+
+#include <stdint.h>
+#include <string.h>
+
+void set_public_addr(void)
+{
+	bt_addr_le_t addr = {BT_ADDR_LE_RANDOM,
+			     {{0x0A, 0x89, 0x67, 0x45, 0x23, 0xC1}}};
+	bt_id_create(&addr, NULL);
+}
+
+void server_procedure(void)
+{
+	struct bt_conn *conn;
+	uint8_t round = get_test_round();
+
+	wait_for_round_start();
+
+	printk("Start test round: %d\n", get_test_round());
+
+	/* Use the same public address for all instances of the central. If we
+	 * don't do that, encryption (using the bond stored in NVS) will
+	 * fail.
+	 */
+	set_public_addr();
+
+	bt_enable(NULL);
+	gatt_register_service_1();
+	if (round >= 2) {
+		gatt_register_service_2();
+	}
+
+	/* Wait for GATT DB hash to complete.
+	 *
+	 * Only then, load the settings. This is done to trigger a bug where the
+	 * hash was overwritten on NVS before the settings were loaded. It's a
+	 * race condition between the loading of the settings and the hash being
+	 * calculated and written to NVS.
+	 */
+	k_sleep(K_SECONDS(2));
+
+	settings_load();
+	/* settings_load could've triggered a new hash calculation. Wait until
+	 * that one too is over.
+	 */
+	k_sleep(K_SECONDS(2));
+
+	conn = connect_as_central();
+	printk("connected: conn %p\n", conn);
+
+	if (get_test_round() == 0) {
+		printk("bonding\n");
+		bond(conn);
+	} else {
+		printk("encrypting\n");
+		set_security(conn, BT_SECURITY_L2);
+	}
+
+	wait_disconnected();
```

Comment:
This doesn't test that the write happens before the disconnection.

Reply:
But also: Does anything bad happen if we power-off in the on-disconnected callback? (I suspect some work items will not run.)

Reply:
I will change that so the server dies before the connection is ended.
For the second point, I think that should rather be in a separate bsim robustness-type test, since I suppose other stuff than GATT might break.